File: src/cpp/cdm/patient/actions/SERespiratoryMechanicsConfiguration.cpp
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

#include "cdm/CommonDefs.h"
#include "cdm/patient/actions/SERespiratoryMechanicsConfiguration.h"
#include "cdm/system/physiology/SERespiratoryMechanics.h"
#include "cdm/io/protobuf/PBPatientActions.h"

SERespiratoryMechanicsConfiguration::SERespiratoryMechanicsConfiguration(Logger* logger) : SEPatientAction(logger)
{
  m_SettingsFile = "";
  m_Settings = nullptr;
  m_AppliedRespiratoryCycle = eAppliedRespiratoryCycle::Expiratory;
  m_MergeType = eMergeType::Append;
}

SERespiratoryMechanicsConfiguration::~SERespiratoryMechanicsConfiguration()
{
  m_SettingsFile = "";
  SAFE_DELETE(m_Settings);
}

void SERespiratoryMechanicsConfiguration::Clear()
{
  SEPatientAction::Clear();
  m_SettingsFile = "";
  if (m_Settings)
    m_Settings->Clear();
  m_AppliedRespiratoryCycle = eAppliedRespiratoryCycle::Expiratory;
  m_MergeType = eMergeType::Append;
}

void SERespiratoryMechanicsConfiguration::Copy(const SERespiratoryMechanicsConfiguration& src, bool /*preserveState*/)
{
  //if(preserveState) // Cache any state before copy,
  PBPatientAction::Copy(src, *this);
  //if(preserveState) // Put back any state
}

bool SERespiratoryMechanicsConfiguration::IsValid() const
{
  return SEPatientAction::IsValid() &&(HasSettings() || HasSettingsFile());
}

bool SERespiratoryMechanicsConfiguration::IsActive() const
{
  return SEPatientAction::IsActive();
}
void SERespiratoryMechanicsConfiguration::Deactivate()
{
  SEPatientAction::Deactivate();
  Clear();//No stateful properties
}

const SEScalar* SERespiratoryMechanicsConfiguration::GetScalar(const std::string& name)
{
  return GetSettings().GetScalar(name);
}

bool SERespiratoryMechanicsConfiguration::HasSettings() const
{
  return m_Settings != nullptr;
}
SERespiratoryMechanics& SERespiratoryMechanicsConfiguration::GetSettings()
{
  if (m_Settings == nullptr)
    m_Settings = new SERespiratoryMechanics(GetLogger());
  return *m_Settings;
}
const SERespiratoryMechanics* SERespiratoryMechanicsConfiguration::GetSettings() const
{
  return m_Settings;
}

std::string SERespiratoryMechanicsConfiguration::GetSettingsFile() const
{
  return m_SettingsFile;
}
void SERespiratoryMechanicsConfiguration::SetSettingsFile(const std::string& fileName)
{
  m_SettingsFile = fileName;
}
bool SERespiratoryMechanicsConfiguration::HasSettingsFile() const
{
  return !m_SettingsFile.empty();
}

void SERespiratoryMechanicsConfiguration::SetAppliedRespiratoryCycle(eAppliedRespiratoryCycle c)
{
  m_AppliedRespiratoryCycle = c;
}
eAppliedRespiratoryCycle SERespiratoryMechanicsConfiguration::GetAppliedRespiratoryCycle() const
{
  return m_AppliedRespiratoryCycle;
}

void SERespiratoryMechanicsConfiguration::SetMergeType(eMergeType m)
{
  m_MergeType = m;
}
eMergeType SERespiratoryMechanicsConfiguration::GetMergeType() const
{
  return m_MergeType;
}

File: src/java/com/kitware/pulse/cdm/patient/actions/SERespiratoryMechanicsConfiguration.java
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

package com.kitware.pulse.cdm.patient.actions;

import com.kitware.pulse.cdm.bind.Actions.eMergeType;
import com.kitware.pulse.cdm.bind.PatientActions.RespiratoryMechanicsConfigurationData;
import com.kitware.pulse.cdm.bind.Physiology.RespiratoryMechanicsData;
import com.kitware.pulse.cdm.system.physiology.SERespiratoryMechanics;

public class SERespiratoryMechanicsConfiguration extends SEPatientAction
  {
    protected SERespiratoryMechanics settings = null;
    protected String settingsFile = "";
    protected eMergeType mergeType = eMergeType.Append;

    public SERespiratoryMechanicsConfiguration()
    {
      super();
    }

    public SERespiratoryMechanicsConfiguration(SERespiratoryMechanicsConfiguration other)
    {
      copy(other);
    }

    public void copy(SERespiratoryMechanicsConfiguration other)
    {
      super.copy(other);
      if (other.settings != null)
        this.getSettings().copy(other.settings);
      this.settingsFile = other.settingsFile;
      mergeType = other.mergeType;
    }

    public void clear()
    {
      super.clear();
      if (this.settings != null)
        this.settings.clear();
      this.settingsFile = "";
      mergeType = eMergeType.Append;
    }
    
    public static void load(RespiratoryMechanicsConfigurationData src, SERespiratoryMechanicsConfiguration dst)
    {
      dst.setMergeType(src.getMergeType());
      if (src.hasSettings())
        SERespiratoryMechanics.load(src.getSettings(), dst.getSettings());
      else if (src.getSettingsFile().isEmpty())
        dst.setSettingsFile(src.getSettingsFile());
    }
    public static RespiratoryMechanicsConfigurationData unload(SERespiratoryMechanicsConfiguration src)
    {
      RespiratoryMechanicsConfigurationData.Builder dst = RespiratoryMechanicsConfigurationData.newBuilder();
      unload(src,dst);
      return dst.build();
    }
    protected static void unload(SERespiratoryMechanicsConfiguration src, RespiratoryMechanicsConfigurationData.Builder dst)
    {
      dst.setMergeType(src.getMergeType());
      if(src.hasSettings())
        dst.setSettings(SERespiratoryMechanics.unload(src.getSettings()));
      else if (src.hasSettingsFile())
        dst.setSettingsFile(src.getSettingsFile());
    }

    public boolean isValid()
    {
      return hasSettings() || hasSettingsFile();
    }

    public boolean hasSettings()
    {
      return this.settings != null;
    }
    public SERespiratoryMechanics getSettings()
    {
      if (this.settings == null)
        this.settings = new SERespiratoryMechanics();
      return this.settings;
    }

    public boolean hasSettingsFile()
    {
      return this.settingsFile != null && !this.settingsFile.isEmpty();
    }
    public String getSettingsFile()
    {
      return this.settingsFile;
    }
    public void setSettingsFile(String s)
    {
      this.settingsFile = s;
    }
    public eMergeType getMergeType()
    {
      return mergeType;
    }
    public void setMergeType(eMergeType m)
    {
      mergeType = m;
    }

    public String toString()
    {
      String str = "Respiratory Mechanics Configuration";
      if (this.hasSettingsFile())
        str += "\n\tSettings File: " + this.settingsFile;
      else if (hasSettings())
      {
        str += settings.toString();
      }
      str += "\n\tMerge Type: " + this.mergeType;
      return str;
    }
  }
